File: csharp/solution_and_tests/freebie/problem_1672_richest_customer_wealth.cs
/*
 * https://leetcode.com/problems/richest-customer-wealth/
 * Explanation:
 * 
 * Straight forward problem, loop through each account then calculate the sum,
 * wealth will be the max of current wealth or the new bank account sum.
 * 
 * Time: O(m * n)
 * Space: O(1)
 */

using System;
using System.Linq;
using NUnit.Framework;

namespace Richest_Customer_Wealth
{
    class Solution
    {
        public static int MaximumWealth(int[][] accounts)
        {
            int richestCustomerWealth = 0;
            foreach (int[] account in accounts)
            {
                richestCustomerWealth = Math.Max(richestCustomerWealth, account.Sum());
            }
            return richestCustomerWealth;
        }
    }

    class Tests
    {
        [Test]
        public void TestMaximumWealth()
        {
            int[][] testCase1 = new int[][] { new int[] { 1, 2, 3 }, new int[] { 3, 2, 1 } };
            int testCase1ExpectedResult = 6;
            int[][] testCase2 = new int[][] { new int[] { 1, 5 }, new int[] { 7, 3 }, new int[] { 3, 5 } };
            int testCase2ExpectedResult = 10;
            int[][] testCase3 = new int[][] { new int[] { 2, 8, 7 }, new int[] { 7, 1, 3 }, new int[] { 1, 9, 5 } };
            int testCase3ExpectedResult = 17;
            int[][] testCase4 = new int[][] { new int[] { 0, 0, 0 }, new int[] { 0, 0, 0 } };
            int testCase4ExpectedResult = 0;
            int[][] testCase5 = new int[][] { new int[] { 5, 5, 5 }, new int[] { 10, 10, 10 }, new int[] { 15, 15, 15 } };
            int testCase5ExpectedResult = 45;
            int[][] testCase6 = new int[][] { new int[] { 1000 } };
            int testCase6ExpectedResult = 1000;

            Assert.AreEqual(testCase1ExpectedResult, Solution.MaximumWealth(testCase1));
            Assert.AreEqual(testCase2ExpectedResult, Solution.MaximumWealth(testCase2));
            Assert.AreEqual(testCase3ExpectedResult, Solution.MaximumWealth(testCase3));
            Assert.AreEqual(testCase4ExpectedResult, Solution.MaximumWealth(testCase4));
            Assert.AreEqual(testCase5ExpectedResult, Solution.MaximumWealth(testCase5));
            Assert.AreEqual(testCase6ExpectedResult, Solution.MaximumWealth(testCase6));
        }
    }
}

File: JavaMaven/src/test/java/freebie/problem_1672_richest_customer_wealth.java
/*
 * https://leetcode.com/problems/richest-customer-wealth/
 * Explanation:
 * 
 * Straight forward problem, loop through each account then calculate the sum,
 * wealth will be the max of current wealth or the new bank account sum.
 * 
 * Time: O(m * n)
 * Space: O(1)
 */

package freebie;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.assertEquals;

public class problem_1672_richest_customer_wealth {
    public int maxWealth(int[][] accounts) {
        int wealth = 0;
        for (int[] account : accounts) {
            int sum = 0;
            for (int num : account) {
                sum += num;
            }
            wealth = Math.max(wealth, sum);
        }
        return wealth;
    }

    @Test
    public void testWealth() {
        int[][] testCase1 = { { 1, 2, 3 }, { 3, 2, 1 } };
        int[][] testCase2 = { { 1, 5 }, { 7, 3 }, { 3, 5 } };
        int[][] testCase3 = { { 2, 8, 7 }, { 7, 1, 3 }, { 1, 9, 5 } };

        assertEquals(6, maxWealth(testCase1));
        assertEquals(10, maxWealth(testCase2));
        assertEquals(17, maxWealth(testCase3));
    }
}